SECP3223-02 DATA ANALYTICS PROGRAMMING

GROUP PROJECT PHASE 1

Group member name:
1. Dayang Farah Farzana binti Abang Idham A23CS0071
2. Farra Nurzahin binti Zaharil Anuar A23CS0079
3. Safiya Nursyahadah binti Masnoor A23CS0176

In [4]:
#We use pandas to load two datasets

import pandas as pd

fuel_df = pd.read_csv("C:/ProjectDAP/FuelConsumption.csv")
cars_df = pd.read_csv("C:/ProjectDAP/cars_2025.csv")
fuel_df
cars_df

,date_reg,type,maker,model,colour,fuel,state
0,2025-01-01,motokar,BYD,Seal,white,electric,Rakan Niaga
1,2025-01-01,window_van,Cam,Placer-X,yellow,greendiesel,Johor
2,2025-01-01,jip,Chery,Jaecoo J7,green,petrol,Rakan Niaga
3,2025-01-01,jip,Chery,Jaecoo J7,silver,petrol,Rakan Niaga
4,2025-01-01,jip,Chery,Tiggo,grey,petrol,Rakan Niaga
...,...,...,...,...,...,...,...
263573,2025-04-30,window_van,Zeekr,009,black,electric,W.P. Kuala Lumpur
263574,2025-04-30,window_van,Zeekr,009,black,electric,W.P. Kuala Lumpur
263575,2025-04-30,window_van,Zeekr,009,black,electric,W.P. Kuala Lumpur
263576,2025-04-30,window_van,Zeekr,009,black,electric,Rakan Niaga


In [6]:
#We examine the structure of the data using;
# 1. (.shape) - shows the number of rows and columns
# 2. (.dtypes) - shows the data tyopes pf each column
# 3. (head()) - displays the first 5 rows of each dataset

print(fuel_df.shape)
print(fuel_df.dtypes)
print(fuel_df.head())

print(cars_df.shape)
print(cars_df.dtypes)
print(cars_df.head())


(1067, 13)
MODELYEAR                     int64
MAKE                         object
MODEL                        object
VEHICLECLASS                 object
ENGINESIZE                  float64
CYLINDERS                     int64
TRANSMISSION                 object
FUELTYPE                     object
FUELCONSUMPTION_CITY        float64
FUELCONSUMPTION_HWY         float64
FUELCONSUMPTION_COMB        float64
FUELCONSUMPTION_COMB_MPG      int64
CO2EMISSIONS                  int64
dtype: object
   MODELYEAR   MAKE       MODEL VEHICLECLASS  ENGINESIZE  CYLINDERS  \
0       2014  ACURA         ILX      COMPACT         2.0          4   
1       2014  ACURA         ILX      COMPACT         2.4          4   
2       2014  ACURA  ILX HYBRID      COMPACT         1.5          4   
3       2014  ACURA     MDX 4WD  SUV - SMALL         3.5          6   
4       2014  ACURA     RDX AWD  SUV - SMALL         3.5          6   

  TRANSMISSION FUELTYPE  FUELCONSUMPTION_CITY  FUELCONSUMPTION_HWY  \
0         

In [8]:
# Print the column names

print(cars_df.columns)
print(fuel_df.columns)

Index(['date_reg', 'type', 'maker', 'model', 'colour', 'fuel', 'state'], dtype='object')
Index(['MODELYEAR', 'MAKE', 'MODEL', 'VEHICLECLASS', 'ENGINESIZE', 'CYLINDERS',
       'TRANSMISSION', 'FUELTYPE', 'FUELCONSUMPTION_CITY',
       'FUELCONSUMPTION_HWY', 'FUELCONSUMPTION_COMB',
       'FUELCONSUMPTION_COMB_MPG', 'CO2EMISSIONS'],
      dtype='object')


In [10]:
#Rename the maker and model columns in the cars_df Dataframe to MAKE and MODEL so that 
#they match the column in fuel_df
cars_df.rename(columns={'maker': 'MAKE', 'model': 'MODEL'}, inplace=True)

# Then clean the new columns:
cars_df['MAKE'] = cars_df['MAKE'].fillna('').str.strip().str.upper()
cars_df['MODEL'] = cars_df['MODEL'].fillna('').str.strip().str.upper()


In [12]:
#Merge both datasets

merged_df = pd.merge(cars_df, fuel_df, how='left', on=['MAKE', 'MODEL'])


In [14]:
#Find and counts all rows in data where the CO2EMISSIONS value is missing

unmatched = merged_df[merged_df['CO2EMISSIONS'].isnull()]
print(f"Unmatched rows: {unmatched.shape[0]}")


Unmatched rows: 251480


In [29]:
#Saves merged data into a CSV file ("Merged_Dataset_Phase1.csv")

merged_df.to_csv("Merged_Dataset_Phase1.csv", index=False)


In [30]:
#Extracts all rows where CO2EMISSIONS is availablle and saves them to a CSV file

matched_df = merged_df[merged_df['CO2EMISSIONS'].notnull()]
matched_df.to_csv("Matched_Dataset_Phase1.csv", index=False)


### 🔍 Matching Summary

- Total car registrations (2025): 263,578
- Total emission records (2014): 1,067
- Total matched rows after merging: ~12,098
- Unmatched rows: 251,480

#### 📌 Observations:
- Many newer or local car models in Malaysia (2025) are not present in the 2014 dataset.
- This leads to a high number of unmatched rows.
- Matching was done using standardized `MAKE` and `MODEL` fields.


In [18]:
#Import pandas for cleaning, preparation and wrangling, from numpy import nan to handle missing value
import pandas as pd
from numpy import nan as NA

#Load the merged dataset
merged_df = pd.read_csv("C:/ProjectDAP/Merged_Dataset_Phase1.csv")

#Display the first few rows to review the data
merged_df.head()

,date_reg,type,MAKE,MODEL,colour,fuel,state,MODELYEAR,VEHICLECLASS,ENGINESIZE,CYLINDERS,TRANSMISSION,FUELTYPE,FUELCONSUMPTION_CITY,FUELCONSUMPTION_HWY,FUELCONSUMPTION_COMB,FUELCONSUMPTION_COMB_MPG,CO2EMISSIONS
0,2025-01-01,motokar,BYD,SEAL,white,electric,Rakan Niaga,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,2025-01-01,window_van,CAM,PLACER-X,yellow,greendiesel,Johor,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,2025-01-01,jip,CHERY,JAECOO J7,green,petrol,Rakan Niaga,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,2025-01-01,jip,CHERY,JAECOO J7,silver,petrol,Rakan Niaga,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,2025-01-01,jip,CHERY,TIGGO,grey,petrol,Rakan Niaga,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [20]:
#Display number of rows before data cleaning
merged_df.shape[0]

277343

In [22]:
#We perform data cleaning and preparation by doing:

#1. Check each column's dtype
for col in merged_df.columns:
    print(col, merged_df[col].dtype)

date_reg object
type object
MAKE object
MODEL object
colour object
fuel object
state object
MODELYEAR float64
VEHICLECLASS object
ENGINESIZE float64
CYLINDERS float64
TRANSMISSION object
FUELTYPE object
FUELCONSUMPTION_CITY float64
FUELCONSUMPTION_HWY float64
FUELCONSUMPTION_COMB float64
FUELCONSUMPTION_COMB_MPG float64
CO2EMISSIONS float64


In [24]:
#2. Classify the same dtype into lists to standardize the text and formats of the file
# We want to have 'MAKE' and 'MODEL' stay in uppercase hence the makemodel_columns 
text_columns = ["date_reg", "type", "colour", "fuel", "state", "VEHICLECLASS",
                "TRANSMISSION", "FUELTYPE"]
numeric_columns = ["MODELYEAR", "ENGINESIZE", "CYLINDERS", "FUELCONSUMPTION_CITY", "FUELCONSUMPTION_HWY", "FUELCONSUMPTION_COMB", 
                   "FUELCONSUMPTION_COMB_MPG", "CO2EMISSIONS"]
makemodel_columns = ["MAKE", "MODEL"]

#3. Fill missing values with 0 or NaN
merged_df[numeric_columns] = merged_df[numeric_columns].fillna(0)

merged_df[text_columns] = merged_df[text_columns].fillna(NA)
merged_df[makemodel_columns] = merged_df[makemodel_columns].fillna(NA)

#4. Remove duplicates rows
merged_df = merged_df.drop_duplicates()
merged_df.shape[0]

53240

In [26]:
#5. Perform:
# str.strip() - remove extra spaces
# Convert all text_columns value to string
# replace word nan to NaN after string cleaning
for col in text_columns:
    merged_df[col] = merged_df[col].astype(str).str.strip()
    merged_df[col] = merged_df[col].replace('nan', 'NaN')

#6. Convert all numeric_columns value to float
for col in numeric_columns:
    merged_df[col] = merged_df[col].astype(float)

merged_df.head()

,date_reg,type,MAKE,MODEL,colour,fuel,state,MODELYEAR,VEHICLECLASS,ENGINESIZE,CYLINDERS,TRANSMISSION,FUELTYPE,FUELCONSUMPTION_CITY,FUELCONSUMPTION_HWY,FUELCONSUMPTION_COMB,FUELCONSUMPTION_COMB_MPG,CO2EMISSIONS
0,2025-01-01,motokar,BYD,SEAL,white,electric,Rakan Niaga,0.0,NaN,0.0,0.0,NaN,NaN,0.0,0.0,0.0,0.0,0.0
1,2025-01-01,window_van,CAM,PLACER-X,yellow,greendiesel,Johor,0.0,NaN,0.0,0.0,NaN,NaN,0.0,0.0,0.0,0.0,0.0
2,2025-01-01,jip,CHERY,JAECOO J7,green,petrol,Rakan Niaga,0.0,NaN,0.0,0.0,NaN,NaN,0.0,0.0,0.0,0.0,0.0
3,2025-01-01,jip,CHERY,JAECOO J7,silver,petrol,Rakan Niaga,0.0,NaN,0.0,0.0,NaN,NaN,0.0,0.0,0.0,0.0,0.0
4,2025-01-01,jip,CHERY,TIGGO,grey,petrol,Rakan Niaga,0.0,NaN,0.0,0.0,NaN,NaN,0.0,0.0,0.0,0.0,0.0


In [28]:
#Show info tocheck datatypes and non-null counts
merged_df.info()

#View the cleaned data
merged_df.head()

<class 'pandas.core.frame.DataFrame'>
Index: 53240 entries, 0 to 277341
Data columns (total 18 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   date_reg                  53240 non-null  object 
 1   type                      53240 non-null  object 
 2   MAKE                      53240 non-null  object 
 3   MODEL                     53240 non-null  object 
 4   colour                    53240 non-null  object 
 5   fuel                      53240 non-null  object 
 6   state                     53240 non-null  object 
 7   MODELYEAR                 53240 non-null  float64
 8   VEHICLECLASS              53240 non-null  object 
 9   ENGINESIZE                53240 non-null  float64
 10  CYLINDERS                 53240 non-null  float64
 11  TRANSMISSION              53240 non-null  object 
 12  FUELTYPE                  53240 non-null  object 
 13  FUELCONSUMPTION_CITY      53240 non-null  float64
 14  FUELCONSUM

,date_reg,type,MAKE,MODEL,colour,fuel,state,MODELYEAR,VEHICLECLASS,ENGINESIZE,CYLINDERS,TRANSMISSION,FUELTYPE,FUELCONSUMPTION_CITY,FUELCONSUMPTION_HWY,FUELCONSUMPTION_COMB,FUELCONSUMPTION_COMB_MPG,CO2EMISSIONS
0,2025-01-01,motokar,BYD,SEAL,white,electric,Rakan Niaga,0.0,NaN,0.0,0.0,NaN,NaN,0.0,0.0,0.0,0.0,0.0
1,2025-01-01,window_van,CAM,PLACER-X,yellow,greendiesel,Johor,0.0,NaN,0.0,0.0,NaN,NaN,0.0,0.0,0.0,0.0,0.0
2,2025-01-01,jip,CHERY,JAECOO J7,green,petrol,Rakan Niaga,0.0,NaN,0.0,0.0,NaN,NaN,0.0,0.0,0.0,0.0,0.0
3,2025-01-01,jip,CHERY,JAECOO J7,silver,petrol,Rakan Niaga,0.0,NaN,0.0,0.0,NaN,NaN,0.0,0.0,0.0,0.0,0.0
4,2025-01-01,jip,CHERY,TIGGO,grey,petrol,Rakan Niaga,0.0,NaN,0.0,0.0,NaN,NaN,0.0,0.0,0.0,0.0,0.0


In [39]:
#Saved the update dataset to a new CSV file
merged_df.to_csv('Cleaned_Merged_Dataset_Phase1.csv', index=False);

### 🔍 Data Cleaning, Preparation and Wrangling Summary

- Total original rows: 277,343
- Total rows after cleaning: 53240
- Total columns: 18 
- Numeric columns cleaned: 8
- Text columns cleaned: 7
- `MAKE` and `MODEL` columns cleaned separately

#### 📌 Observations:
- All missing numeric values filled with 0 while missing text values replaced with 'NaN'
- Text values were standardized by stripped extra spaces and converted to string type
- Duplicate rows were removed

In [40]:
# Group the data by MAKE and MODEL
agg_df = (
    fuel_df.groupby(['MAKE', 'MODEL'])
    .agg(
        count=('MODEL', 'count'),  # Count of records
        avg_fuel_consumption=('FUELCONSUMPTION_COMB', 'mean'),  # Average combined fuel use
        avg_co2_emission=('CO2EMISSIONS', 'mean')  # Average CO2 emission
    )
    .reset_index()
    .sort_values(by='count', ascending=False) 
)

# Display the aggregated DataFrame
print(agg_df.head())

# Save the final dataset to CSV for Phase II
agg_df.to_csv("aggregated_fuel_data.csv", index=False)

           MAKE         MODEL  count  avg_fuel_consumption  avg_co2_emission
229        FORD      F150 FFV      8             15.775000        299.500000
230        FORD  F150 FFV 4X4      8             16.550000        314.125000
274       HONDA        ACCORD      6              9.050000        208.166667
239        FORD     FOCUS FFV      6              9.116667        173.333333
636  VOLKSWAGEN        BEETLE      6              9.183333        211.166667


📌 Aggregation & Final Dataset Preparation
This section focuses on aggregating key metrics from the FuelConsumption.csv dataset to prepare a summarized dataset for further analysis in Phase II.

Steps performed : 

- Data was grouped by MAKE and MODEL to consolidate vehicle entries.
- Three main aggregations were performed for each group :
  1. count : Number of cars for each MAKE-MODEL combination.
  2. avg_fuel_consumption : Average combined fuel consumption.
  3. avg_co2_emission : Average CO2 emission.

The grouped data was sorted by the number of records to highlight the most common car models.

The final cleaned and aggregated dataset was saved as aggregated_fuel_data.csv for use in next phase.